### PR TITLE
fix: Do NOT require list permissions on a single set of tags

### PR DESCRIPTION
**Pull request by ivanilves**
closes https://github.com/ivanilves/lstags/issues/222

### CODE REVIEW

File: api/v1/registry/client/client.go [hunk 1]
```diff
@@ -172,8 +172,6 @@ func decodeTagData(body io.ReadCloser) ([]string, map[string]manifest.Manifest,
 		return nil, nil, err
 	}
 
-	tagManifests = manifest.MapByTag(tagManifests)
```
Comment (ivanilves) [inline api/v1/registry/client/client.go:175]: this was a useless line :laughing:

File: repository/repository.go [hunk 2]
```diff
@@ -263,6 +263,7 @@ func ParseRef(ref string) (*Repository, error) {
 		refParts := strings.Split(fullRef, "=")
 		fullRepo = refParts[0]
 		repoTags = strings.Split(refParts[1], ",")
+		isSingle = true
```
Comment (ivanilves) [inline repository/repository.go:266]: a set of single tags should be also classified as "single"

File: api/v1/registry/client/client.go [hunk 3]
```diff
@@ -248,6 +259,17 @@ func (cli *RegistryClient) TagData(repoPath string) ([]string, map[string]manife
 	return allTagNames, allTagManifests, nil
 }
 
+// SingleTagData gets data for a set of single tags
+func (cli *RegistryClient) SingleTagData(repoTags []string) ([]string, map[string]manifest.Manifest, error) {
+	tagManifests := make(map[string]manifest.Manifest)
+
+	for _, tagName := range repoTags {
+		tagManifests[tagName] = manifest.Manifest{}
```
Comment (ivanilves) [inline api/v1/registry/client/client.go:267]: this would be filled later...

**Comment by cablespaghetti** [comment]
Looks cool! I'll try and build this and try it out :D

**Comment by cablespaghetti** [comment]
This is SO MUCH FASTER :grin:

**Review by cablespaghetti — APPROVED** [review]

**Comment by cablespaghetti** [comment]
I may have spoken too soon. I think this behaves differently with a config file maybe? If I run a single instance locally it's lightning fast but when run in my environment I get this still:

image

**Comment by ivanilves** [comment]
First, thank you for the swift review @cablespaghetti :bow: 

> I may have spoken too soon. I think this behaves differently with a config file maybe? If I run a single instance locally it's lightning fast but when run in my environment I get this still:

I think this may be related to `quay.io` API rate limits, the infamous yet frequent **error 429**.

Original issue cases were **error 401** because of list requests instead of "blind" get ones :slightly_smiling_face: 

Could you please add concurrency limits for quay:
* `-c 1` - to disable concurrency (this may result in considerable slow down, just asking to test to see if it works);
* ... or `-w 50ms` to throttle requests while keeping concurrency enabled (curious, how it will slow down and will it work for you);

:question: 

```
  -c, --concurrent-requests=  Limit of concurrent requests to the registry (default: 16) [$CONCURRENT_REQUESTS]
  -w, --wait-between=         Time to wait between batches of requests (incl. pulls and pushes) (default: 0) [$WAIT_BETWEEN]
```

P.S.:
`quay.io` is pretty sensible/touchy on API rate limiting, I even think to change default concurrency setting for `quay.io` requests...

**Review by vonrabbe — APPROVED** [review]

**Comment by vonrabbe** [comment]
I agree @ivanilves this "429 Too Many Requests" is a separate issue :+1:

**Pull request merged**